client/alloc_runner_health_watcher.go

```diff
@@ -14,220 +19,481 @@ const (
 	// consulCheckLookupInterval is the  interval at which we check if the
 	// Consul checks are healthy or unhealthy.
 	consulCheckLookupInterval = 500 * time.Millisecond
+
+	// allocHealthEventSource is the source used for emitting task events
+	allocHealthEventSource = "Alloc Unhealthy"
 )
 
 // watchHealth is responsible for watching an allocation's task status and
-// potentially consul health check status to determine if the allocation is
+// potentially Consul health check status to determine if the allocation is
 // healthy or unhealthy.
 func (r *AllocRunner) watchHealth(ctx context.Context) {
+
 	// See if we should watch the allocs health
 	alloc := r.Alloc()
-	if alloc.DeploymentID == "" {
-		r.logger.Printf("[TRACE] client.alloc_watcher: exiting because alloc isn't part of a deployment")
-		return
-	} else if alloc.DeploymentStatus.IsHealthy() || alloc.DeploymentStatus.IsUnhealthy() {
-		r.logger.Printf("[TRACE] client.alloc_watcher: exiting because alloc deployment health already determined")
+	if alloc.DeploymentID == "" || alloc.DeploymentStatus.IsHealthy() || alloc.DeploymentStatus.IsUnhealthy() {
 		return
 	}
 
 	tg := alloc.Job.LookupTaskGroup(alloc.TaskGroup)
 	if tg == nil {
 		r.logger.Printf("[ERR] client.alloc_watcher: failed to lookup allocation's task group. Exiting watcher")
 		return
+	} else if tg.Update == nil || tg.Update.HealthCheck == structs.UpdateStrategyHealthCheck_Manual {
+		return
 	}
 
-	// Checks marks whether we should be watching for Consul health checks
-	desiredChecks := 0
-	var checkTicker *time.Ticker
-	var checkCh <-chan time.Time
+	// Get an allocation listener to watch for alloc events
+	l := r.allocBroadcast.Listen()
+	defer l.Close()
 
-	u := tg.Update
-	switch {
-	case u == nil:
-		r.logger.Printf("[TRACE] client.alloc_watcher: no update block for alloc %q. exiting", alloc.ID)
-		return
-	case u.HealthCheck == structs.UpdateStrategyHealthCheck_Manual:
-		r.logger.Printf("[TRACE] client.alloc_watcher: update block has manual checks for alloc %q. exiting", alloc.ID)
-		return
-	case u.HealthCheck == structs.UpdateStrategyHealthCheck_Checks:
-		for _, task := range tg.Tasks {
-			for _, s := range task.Services {
-				desiredChecks += len(s.Checks)
-			}
+	// Create a new context with the health deadline
+	deadline := time.Now().Add(tg.Update.HealthyDeadline)
+	healthCtx, healthCtxCancel := context.WithDeadline(ctx, deadline)
+	defer healthCtxCancel()
+	r.logger.Printf("[DEBUG] client.alloc_watcher: deadline (%v) for alloc %q is at %v", tg.Update.HealthyDeadline, alloc.ID, deadline)
+
+	// Create the health tracker object
+	tracker := newAllocHealthTracker(healthCtx, r.logger, alloc, l, r.consulClient)
+	tracker.Start()
+
+	allocHealthy := false
+	select {
+	case <-healthCtx.Done():
+		// We were cancelled which means we are no longer needed
+		if healthCtx.Err() == context.Canceled {
+			return
 		}
 
-		checkTicker = time.NewTicker(consulCheckLookupInterval)
-		checkCh = checkTicker.C
+		break
```

Comment:
No need for the `break` but maybe add a comment mentioning this is the "not-healthy-by-deadline" case.